packages/react/src/Pagination/Pagination.tsx

```diff
@@ -0,0 +1,192 @@
+  /**
+   * Callback triggered when interaction changes the page number.
+   */
+  // eslint-disable-next-line no-unused-vars
+  onPageChange?: (page: number) => void
+  /**
+   * The current page number.
+   */
+  page?: number
+  /**
+   * The number of items per page.
+   */
+  pageSize?: number
+  /**
+   * The maximum amount of pages shown. This has a lower limit of 5
+   */
+  maxVisiblePages?: number
+}
+
+/**
+ * This returns an array of the range, including spacers
+ *
+ * @example
+ * currentPage = 4, totalPages = 7, maxVisiblePages = 7
+ * // returns [1, 2, 3, 4, 5, 6, 7]
+ *
+ * @example
+ * currentPage = 5, totalPages = 100, maxVisiblePages = 7
+ * // returns [1, 'firstSpacer', 4, 5, 6, 'lastSpacer', 100]
+ *
+ * @example
+ * currentPage = 97, totalPages = 100, maxVisiblePages = 7
+ * // returns [1, 'firstSpacer', 96, 97, 98, 99, 100]
+ */
+
+function getRange(currentPage: number, totalPages: number, maxVisiblePages: number): Array<string | number> {
+  // the total amount of visible pages is whatever's lower, totalPages or maxVisiblePages
+  // maxVisiblePages has a lower limit of 5
+  const visiblePages = Math.min(totalPages, Math.max(maxVisiblePages, 5))
+
+  const min = 1
+  // the center part of the range starts with the current page minus half of the visible pages
+  let centerStartPage = currentPage - Math.floor(visiblePages / 2)
+  // centerStartPage has a lower limit of 1
+  centerStartPage = Math.max(centerStartPage, min)
+  // centerStartPage has an upper limit of 1 plus total pages minus visible pages
+  centerStartPage = Math.min(centerStartPage, min + totalPages - visiblePages)
+
+  const pages = Array.from({ length: visiblePages }, (_el, i) => centerStartPage + i).reduce<Array<string | number>>(
+    (acc, pageNr, index) => {
+      if (index === 0 && pageNr !== 1) {
+        return [1, 'firstSpacer']
+      }
+
+      if (totalPages > visiblePages && index === visiblePages - 2 && currentPage < totalPages - 2) {
+        return [...acc, 'lastSpacer', totalPages]
+      }
+      // Skip a number when spacer is already added
+      if ((acc.includes('firstSpacer') && index === 1) || (acc.includes('lastSpacer') && index === visiblePages - 1)) {
+        return acc
+      }
+      return [...acc, pageNr]
+    },
+    [],
+  )
+
+  return pages
+}
+
+export const Pagination = forwardRef(
+  (
+    {
+      className,
+      collectionSize,
+      onPageChange,
+      page = 1,
+      pageSize = 10,
+      maxVisiblePages = 7,
+      ...restProps
+    }: PaginationProps,
+    ref: ForwardedRef<HTMLElement>,
+  ) => {
+    const [currentPage, setCurrentPage] = useState(page)
+
+    // If page prop changes, set currentPage so component can be managed externally
+    useEffect(() => {
+      setCurrentPage(page)
+    }, [page])
+
+    const totalPages = Math.ceil(collectionSize / pageSize)
+
+    // Get array of page numbers and / or spacers
+    const range = getRange(currentPage, totalPages, maxVisiblePages)
```

Comment:
For a small performance improvement would it be better to memo this? Every state change outside of the Pagination component would run this function even nothing happens here.